// @ant-design/x-markdown/es/XMarkdown/__benchmark__/components/MarkdownRenderer.js
import MarkdownIt from 'markdown-it';
// @ts-ignore - benchmark only, ignore type checking
import markdownItKatex from 'markdown-it-katex';
import { Marked } from 'marked';
import React from 'react';
import ReactMarkdown from 'react-markdown';
import rehypeKatex from 'rehype-katex';
import rehypeRaw from 'rehype-raw';
import remarkGfm from 'remark-gfm';
import remarkMath from 'remark-math';
import { Streamdown } from 'streamdown';
import getLatex from "../../../plugins/Latex";
import XMarkdown from "../../index";
const md = new MarkdownIt();
// benchmark only: bypass TS check
md.use(markdownItKatex);
const marked = new Marked({
  extensions: getLatex()
});
const MarkedRenderer = props => /*#__PURE__*/React.createElement("div", {
  className: "markdown-container"
  // biome-ignore lint/security/noDangerouslySetInnerHtml: benchmark only
  ,
  dangerouslySetInnerHTML: {
    __html: marked.parse(props.md)
  }
});
const MarkdownItRenderer = props => {
  return (
    /*#__PURE__*/
    // biome-ignore lint/security/noDangerouslySetInnerHtml: benchmark only
    React.createElement("div", {
      className: "markdown-container",
      dangerouslySetInnerHTML: {
        __html: md.render(props.md)
      }
    })
  );
};
const ReactMarkdownRenderer = props => /*#__PURE__*/React.createElement("div", {
  className: "markdown-container"
}, /*#__PURE__*/React.createElement(ReactMarkdown, {
  rehypePlugins: [rehypeRaw, rehypeKatex],
  remarkPlugins: [remarkGfm, remarkMath]
}, props.md));
const XMarkdownRenderer = props => /*#__PURE__*/React.createElement("div", {
  className: "markdown-container"
}, /*#__PURE__*/React.createElement(XMarkdown, {
  streaming: {
    hasNextChunk: props?.hasNextChunk,
    enableAnimation: true
  },
  config: {
    extensions: getLatex()
  }
}, props.md));
const StreamdownRenderer = props => /*#__PURE__*/React.createElement("div", {
  className: "markdown-container"
}, /*#__PURE__*/React.createElement(Streamdown, {
  rehypePlugins: [rehypeRaw, rehypeKatex],
  remarkPlugins: [remarkGfm, remarkMath]
}, props.md));
const Empty = () => /*#__PURE__*/React.createElement("div", null);
export { MarkedRenderer, MarkdownItRenderer, ReactMarkdownRenderer, XMarkdownRenderer, StreamdownRenderer, Empty };
