// @ant-design/x/lib/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Actions", {
  enumerable: true,
  get: function () {
    return _actions.default;
  }
});
Object.defineProperty(exports, "Attachments", {
  enumerable: true,
  get: function () {
    return _attachments.default;
  }
});
Object.defineProperty(exports, "Bubble", {
  enumerable: true,
  get: function () {
    return _bubble.default;
  }
});
Object.defineProperty(exports, "CodeHighlighter", {
  enumerable: true,
  get: function () {
    return _codeHighlighter.default;
  }
});
Object.defineProperty(exports, "Conversations", {
  enumerable: true,
  get: function () {
    return _conversations.default;
  }
});
Object.defineProperty(exports, "FileCard", {
  enumerable: true,
  get: function () {
    return _fileCard.default;
  }
});
Object.defineProperty(exports, "Mermaid", {
  enumerable: true,
  get: function () {
    return _mermaid.default;
  }
});
Object.defineProperty(exports, "Prompts", {
  enumerable: true,
  get: function () {
    return _prompts.default;
  }
});
Object.defineProperty(exports, "Sender", {
  enumerable: true,
  get: function () {
    return _sender.default;
  }
});
Object.defineProperty(exports, "SenderSwitch", {
  enumerable: true,
  get: function () {
    return _SenderSwitch.default;
  }
});
Object.defineProperty(exports, "Sources", {
  enumerable: true,
  get: function () {
    return _sources.default;
  }
});
Object.defineProperty(exports, "Suggestion", {
  enumerable: true,
  get: function () {
    return _suggestion.default;
  }
});
Object.defineProperty(exports, "Think", {
  enumerable: true,
  get: function () {
    return _think.default;
  }
});
Object.defineProperty(exports, "ThoughtChain", {
  enumerable: true,
  get: function () {
    return _thoughtChain.default;
  }
});
Object.defineProperty(exports, "Welcome", {
  enumerable: true,
  get: function () {
    return _welcome.default;
  }
});
Object.defineProperty(exports, "XProvider", {
  enumerable: true,
  get: function () {
    return _xProvider.default;
  }
});
Object.defineProperty(exports, "notification", {
  enumerable: true,
  get: function () {
    return _notification.default;
  }
});
Object.defineProperty(exports, "version", {
  enumerable: true,
  get: function () {
    return _version.default;
  }
});
var _actions = _interopRequireDefault(require("./actions"));
var _attachments = _interopRequireDefault(require("./attachments"));
var _bubble = _interopRequireDefault(require("./bubble"));
var _codeHighlighter = _interopRequireDefault(require("./code-highlighter"));
var _conversations = _interopRequireDefault(require("./conversations"));
var _fileCard = _interopRequireDefault(require("./file-card"));
var _mermaid = _interopRequireDefault(require("./mermaid"));
var _notification = _interopRequireDefault(require("./notification"));
var _prompts = _interopRequireDefault(require("./prompts"));
var _sender = _interopRequireDefault(require("./sender"));
var _SenderSwitch = _interopRequireDefault(require("./sender/SenderSwitch"));
var _sources = _interopRequireDefault(require("./sources"));
var _suggestion = _interopRequireDefault(require("./suggestion"));
var _think = _interopRequireDefault(require("./think"));
var _thoughtChain = _interopRequireDefault(require("./thought-chain"));
var _version = _interopRequireDefault(require("./version"));
var _welcome = _interopRequireDefault(require("./welcome"));
var _xProvider = _interopRequireDefault(require("./x-provider"));
