// @ant-design/x/es/sender/components/LoadingButton/StopLoading.js
import React, { memo } from 'react';
import { useLocale } from "../../../locale";
import enUS from "../../../locale/en_US";
const StopLoadingIcon = /*#__PURE__*/memo(props => {
  const {
    className
  } = props;
  const [contextLocale] = useLocale('Sender', enUS.Sender);
  return /*#__PURE__*/React.createElement("svg", {
    color: "currentColor",
    viewBox: "0 0 1000 1000",
    xmlns: "http://www.w3.org/2000/svg",
    className: className
  }, /*#__PURE__*/React.createElement("title", null, contextLocale.stopLoading), /*#__PURE__*/React.createElement("rect", {
    fill: "currentColor",
    height: "250",
    rx: "24",
    ry: "24",
    width: "250",
    x: "375",
    y: "375"
  }), /*#__PURE__*/React.createElement("circle", {
    cx: "500",
    cy: "500",
    fill: "none",
    r: "450",
    stroke: "currentColor",
    strokeWidth: "100",
    opacity: "0.45"
  }), /*#__PURE__*/React.createElement("circle", {
    cx: "500",
    cy: "500",
    fill: "none",
    r: "450",
    stroke: "currentColor",
    strokeWidth: "100",
    strokeDasharray: "600 9999999"
  }, /*#__PURE__*/React.createElement("animateTransform", {
    attributeName: "transform",
    dur: "1s",
    from: "0 500 500",
    repeatCount: "indefinite",
    to: "360 500 500",
    type: "rotate"
  })));
});
export default StopLoadingIcon;
