// @ant-design/x-sdk/es/x-conversations/index.js
import { useEffect, useState, useSyncExternalStore } from 'react';
import { ConversationStore } from "./store";
export default function useXConversations(config) {
  const [store] = useState(() => {
    const store = new ConversationStore(config?.defaultConversations || [], config?.defaultActiveConversationKey || '');
    return store;
  });
  useEffect(() => {
    return () => {
      store.destroy();
    };
  }, []);
  const conversations = useSyncExternalStore(store.subscribe, store.getSnapshot, store.getSnapshot);
  const activeConversationKey = useSyncExternalStore(store.subscribe, store.getActiveConversationKey, store.getActiveConversationKey);
  return {
    conversations,
    activeConversationKey: activeConversationKey,
    setActiveConversationKey: store.setActiveConversationKey,
    addConversation: store.addConversation,
    removeConversation: store.removeConversation,
    setConversation: store.setConversation,
    getConversation: store.getConversation,
    setConversations: store.setConversations,
    getMessages: store.getMessages
  };
}
