// @ant-design/x-markdown/lib/XMarkdown/__benchmark__/components/MarkdownRenderer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.XMarkdownRenderer = exports.StreamdownRenderer = exports.ReactMarkdownRenderer = exports.MarkedRenderer = exports.MarkdownItRenderer = exports.Empty = void 0;
var _markdownIt = _interopRequireDefault(require("markdown-it"));
var _markdownItKatex = _interopRequireDefault(require("markdown-it-katex"));
var _marked = require("marked");
var _react = _interopRequireDefault(require("react"));
var _reactMarkdown = _interopRequireDefault(require("react-markdown"));
var _rehypeKatex = _interopRequireDefault(require("rehype-katex"));
var _rehypeRaw = _interopRequireDefault(require("rehype-raw"));
var _remarkGfm = _interopRequireDefault(require("remark-gfm"));
var _remarkMath = _interopRequireDefault(require("remark-math"));
var _streamdown = require("streamdown");
var _Latex = _interopRequireDefault(require("../../../plugins/Latex"));
var _index = _interopRequireDefault(require("../../index"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
// @ts-ignore - benchmark only, ignore type checking

const md = new _markdownIt.default();
// benchmark only: bypass TS check
md.use(_markdownItKatex.default);
const marked = new _marked.Marked({
  extensions: (0, _Latex.default)()
});
const MarkedRenderer = props => /*#__PURE__*/_react.default.createElement("div", {
  className: "markdown-container"
  // biome-ignore lint/security/noDangerouslySetInnerHtml: benchmark only
  ,
  dangerouslySetInnerHTML: {
    __html: marked.parse(props.md)
  }
});
exports.MarkedRenderer = MarkedRenderer;
const MarkdownItRenderer = props => {
  return (
    /*#__PURE__*/
    // biome-ignore lint/security/noDangerouslySetInnerHtml: benchmark only
    _react.default.createElement("div", {
      className: "markdown-container",
      dangerouslySetInnerHTML: {
        __html: md.render(props.md)
      }
    })
  );
};
exports.MarkdownItRenderer = MarkdownItRenderer;
const ReactMarkdownRenderer = props => /*#__PURE__*/_react.default.createElement("div", {
  className: "markdown-container"
}, /*#__PURE__*/_react.default.createElement(_reactMarkdown.default, {
  rehypePlugins: [_rehypeRaw.default, _rehypeKatex.default],
  remarkPlugins: [_remarkGfm.default, _remarkMath.default]
}, props.md));
exports.ReactMarkdownRenderer = ReactMarkdownRenderer;
const XMarkdownRenderer = props => /*#__PURE__*/_react.default.createElement("div", {
  className: "markdown-container"
}, /*#__PURE__*/_react.default.createElement(_index.default, {
  streaming: {
    hasNextChunk: props?.hasNextChunk,
    enableAnimation: true
  },
  config: {
    extensions: (0, _Latex.default)()
  }
}, props.md));
exports.XMarkdownRenderer = XMarkdownRenderer;
const StreamdownRenderer = props => /*#__PURE__*/_react.default.createElement("div", {
  className: "markdown-container"
}, /*#__PURE__*/_react.default.createElement(_streamdown.Streamdown, {
  rehypePlugins: [_rehypeRaw.default, _rehypeKatex.default],
  remarkPlugins: [_remarkGfm.default, _remarkMath.default]
}, props.md));
exports.StreamdownRenderer = StreamdownRenderer;
const Empty = () => /*#__PURE__*/_react.default.createElement("div", null);
exports.Empty = Empty;
